sdk/eventhubs/azure_messaging_eventhubs/src/common/connection_manager.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All Rights reserved
+// Licensed under the MIT license.
+
+use crate::common::user_agent::get_package_version;
+use crate::{
+    common::user_agent::{get_package_name, get_platform_info, get_user_agent},
+    error::{ErrorKind, EventHubsError},
+    models::AmqpValue,
+};
+use azure_core::credentials::AccessToken;
+use azure_core::Result;
+use azure_core_amqp::{
+    AmqpClaimsBasedSecurity, AmqpClaimsBasedSecurityApis as _, AmqpConnection,
+    AmqpConnectionApis as _, AmqpConnectionOptions, AmqpSession, AmqpSessionApis as _, AmqpSymbol,
+};
+use std::collections::HashMap;
+use std::sync::{Arc, OnceLock};
+use tokio::sync::Mutex;
+use tracing::{debug, trace};
+use url::Url;
+
+/// The connection manager is responsible for managing the connection to the Event Hubs service.
+/// It also handles authorization and connection recovery.
+///
+/// Currently the connection manager only handles a *single* connection, eventually it will manage
+/// a pool of connections to the service.
+pub(crate) struct ConnectionManager {
+    url: Url,
+    application_id: Option<String>,
+    custom_endpoint: Option<Url>,
+    connections: Mutex<OnceLock<Arc<AmqpConnection>>>,
+    authorization_scopes: Mutex<HashMap<Url, AccessToken>>,
+}
+
+impl ConnectionManager {
+    pub fn new(url: Url, application_id: Option<String>, custom_endpoint: Option<Url>) -> Self {
+        Self {
+            url,
+            application_id,
+            custom_endpoint,
+            connections: Mutex::new(OnceLock::new()),
+            authorization_scopes: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub(crate) async fn ensure_connection(&self) -> Result<()> {
+        let connections = self.connections.lock().await;
+        if connections.get().is_some() {
+            return Ok(());
+        }
+
+        trace!("Creating connection for {}.", self.url);
+        let connection = Arc::new(AmqpConnection::new());
+        connection
+            .open(
+                self.application_id
+                    .clone()
+                    .unwrap_or(uuid::Uuid::new_v4().to_string()),
```

Comment:
Again, this eagerly allocates to `unwrap_or_else` is generally better. Probably only use `unwrap_or` for primitives.